### DISCUSSION TITLE
issue: isDirty not updating with field array/controlled input

### DISCUSSION THREAD

**Discussion by cameron-burris** [react-hook-form/react-hook-form#10796] [Q&A] [answered]
### Version Number

7.43.9

### Codesandbox/Expo snack

https://codesandbox.io/s/modest-lucy-cq2y7j?file=/src/App.tsx

### Steps to reproduce

1. Modify input value
2. Click submit

### Expected behaviour

Should see in console output that formState.isDirty is true and formState.dirtyFields has dirtied input.

You will see that isDirty and dirtyFields will be false and not have the modified input. If you modify the input again and then click submit again, you will see isDirty and dirtyFields is updated to be true and have the modified input.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
<img width="1103" alt="Screenshot 2023-08-10 at 7 59 52 am" src="https://github.com/react-hook-form/react-hook-form/assets/10513364/8a8dfb3e-3e5d-4935-b1a5-65a21db79db0">

working as expected, and the form state is on each render.

**Comment by cameron-burris** [comment]
Alright, I understand now why the formState isn't updating since it is wrapped in a proxy and nothing is using formState during render. Is there a way to get an updated formState on the onSubmit callback without having to use formState somewhere in the render? I would like to do some logic based on if certain fields are dirty or not during submit. At the moment I have a workaround of manually calling form.trigger() to update the formState. This seems to work but not sure if that is a good solution.

**Reply by bluebill1049** [reply, marked as answer]
We did formState with proxy to prevent re-render, if a user is not subscribed to the particular formState. `form.trigger()` that's fine, but it would validate the entire form. (if you concerns about the perf)
